Library/Homebrew/cmd/list.rb

```diff
@@ -1,233 +1,242 @@
+            list_casks if casks.present?
+          end
+        end
       end
-      if args.cask? || (!args.formula? && args.no_named?)
-        cask_names = if args.no_named?
-          Cask::Caskroom.casks
+
+      private
+
+      def filtered_list
+        names = if args.no_named?
+          Formula.racks
         else
-          args.named.to_formulae_and_casks(only: :cask, method: :resolve)
+          racks = args.named.map { |n| Formulary.to_rack(n) }
+          racks.select do |rack|
+            Homebrew.failed = true unless rack.exist?
+            rack.exist?
+          end
         end
-        full_cask_names = cask_names.map(&:full_name).sort(&tap_and_name_comparison)
-        full_cask_names = Formatter.columns(full_cask_names) unless args.public_send(:"1?")
-        puts full_cask_names if full_cask_names.present?
-      end
-    elsif args.pinned?
-      filtered_list(args:)
-    elsif args.versions?
-      filtered_list(args:) unless args.cask?
-      list_casks(args:) if args.cask? || (!args.formula? && !args.multiple? && args.no_named?)
-    elsif args.no_named?
-      ENV["CLICOLOR"] = nil
-
-      ls_args = []
-      ls_args << "-1" if args.public_send(:"1?")
-      ls_args << "-l" if args.l?
-      ls_args << "-r" if args.r?
-      ls_args << "-t" if args.t?
-
-      if !args.cask? && HOMEBREW_CELLAR.exist? && HOMEBREW_CELLAR.children.any?
-        ohai "Formulae" if $stdout.tty? && !args.formula?
-        safe_system "ls", *ls_args, HOMEBREW_CELLAR
-        puts if $stdout.tty? && !args.formula?
-      end
-      if !args.formula? && Cask::Caskroom.any_casks_installed?
-        ohai "Casks" if $stdout.tty? && !args.cask?
-        safe_system "ls", *ls_args, Cask::Caskroom.path
-      end
-    else
-      kegs, casks = args.named.to_kegs_to_casks
-
-      if args.verbose? || !$stdout.tty?
-        find_args = %w[-not -type d -not -name .DS_Store -print]
-        system_command! "find", args: kegs.map(&:to_s) + find_args, print_stdout: true if kegs.present?
-        system_command! "find", args: casks.map(&:caskroom_path) + find_args, print_stdout: true if casks.present?
-      else
-        kegs.each { |keg| PrettyListing.new keg } if kegs.present?
-        list_casks(args:) if casks.present?
-      end
-    end
-  end
+        if args[:pinned?]
+          pinned_versions = {}
+          names.sort.each do |d|
+            keg_pin = (HOMEBREW_PINNED_KEGS/d.basename.to_s)
+            pinned_versions[d] = keg_pin.readlink.basename.to_s if keg_pin.exist? || keg_pin.symlink?
+          end
+          pinned_versions.each do |d, version|
+            puts d.basename.to_s.concat(args[:versions?] ? " #{version}" : "")
+          end
+        else # --versions without --pinned
+          names.sort.each do |d|
+            versions = d.subdirs.map { |pn| pn.basename.to_s }
+            next if args[:multiple?] && versions.length < 2
 
-  def self.filtered_list(args:)
```

Comment:
I just realized that we no longer need to thread `args` through other methods (it's now an ivar, since each command is instance), so I've simplified some method signatures accordingly.